### PR TITLE
[WIP] Checkstyle based report

### PR DESCRIPTION
**Pull request by dmolineus**
This is the WIP pull request for #31 containing following changes

- [x] A generic `Report` class which collect all generated output. I'd like to use it also to collect other output, like generated images, summaries etc.
- [x] It generates a checkstyle formatted output
- [x] It extends the task factory to create post processors
- [x] It contains a checkstyle post process which imports existing post style files
- [x] Plugin API is not changed yet, if we agree on the new API it will be adopted
- [x] Creating named temporary files needs to be implemented. A tool may require a name pattern (e.g. psalm requires a `.xml` file)
- [x] How should we handle tools which does not create files? Should the post processor get the console output of another task?
- [x] Keep running should be used all the time
- [x] Console output should be created based on the report

Adding working bootstrap example for psalm would be:

```php
<?php

use Phpcq\PluginApi\Version10\BuildConfigInterface;
use Phpcq\PluginApi\Version10\ConfigurationOptionsBuilderInterface;
use Phpcq\PluginApi\Version10\ConfigurationPluginInterface;

return new class implements ConfigurationPluginInterface {
    public function getName(): string
    {
        return 'psalm';
    }

    public function describeOptions(ConfigurationOptionsBuilderInterface $configOptionsBuilder): void
    {
        $configOptionsBuilder
            ->describeBoolOption('debug', 'Show debug information.')
            ->describeBoolOption('debug_by_line', 'Debug information on a line-by-line level')
            ->describeBoolOption('shepherd', 'Send data to Shepherd, Psalm\'s GitHub integration tool.')
            ->describeStringOption('shepherd_host', 'Override shepherd host');

        $configOptionsBuilder->describeStringOption(
            'custom_flags',
            'Any custom flags to pass to phpunit. For valid flags refer to the phpunit documentation.',
        );
    }

    public function processConfig(array $config, BuildConfigInterface $buildConfig): iterable
    {
        // Fixme: We need a proper way to create temp files
        $tmpfile = $buildConfig->getBuildTempDir() . '/psalm.checkstyle.xml';

        yield $buildConfig
            ->getTaskFactory()
            ->buildRunPhar('psalm', $this->buildArguments($config, $tmpfile))
            ->withWorkingDirectory($buildConfig->getProjectConfiguration()->getProjectRootPath())
            ->build();

        yield $buildConfig
            ->getTaskFactory()
            ->buildCheckStyleFilePostProcessor('psalm', $tmpfile);
    }

    private function buildArguments(array $config, string $tempFile): array
    {
        $arguments = [];

        foreach (['debug', 'debug_by_line'] as $flag) {
            if (isset($config[$flag])) {
                $arguments[] = '--' .  str_replace('_', '-', $flag);
            }
        }

        if (isset($config['shepherd'])) {
            if (isset($config['shepherd_host'])) {
                $arguments[] = '--shepherd=' . $config['shepherd_host'];
            } else {
                $arguments[] = '--shepherd';
            }
        }

        if ('' !== ($values = $config['custom_flags'] ?? '')) {
            $arguments[] = $values;
        }

        $arguments[] = '--report=' . $tempFile;

        return $arguments;
    }
};
``` 

/cc @discordier

### CODE REVIEW

**Comment by dmolineus** [comment]
I revised the implementation and the concept is ready to be discussed.

1. PHPCQ will create an own xml report file containing more information which would fit into checkstyle. 
2. A checkstyle file is generated as separate file so that other tools may process it.
3. The post processor is not a custom task anymore but bound to the task runner. 
This allows to convert console output for tools which does not support a file output
4. If no custom post processor is registered the console output is used

File: composer.json [hunk 1]
```diff
@@ -23,16 +23,18 @@
...
+        "ext-dom": "*",
         "ext-hash": "*",
         "ext-json": "*",
         "ext-xml": "*",
         "composer/semver": "^1.5",
         "composer/xdebug-handler": "^1.4",
         "guzzlehttp/guzzle": "^6.0",
         "phpcq/gnupg": "^1.0@dev",
-        "phpcq/plugin-api": "^1.0.0@dev",
+        "phpcq/plugin-api": "dev-feature/checkstyle",
```
Comment (dmolineus) [inline composer.json:34]: This has to be changed before merging

File: doc/report.xsd [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+      </xs:documentation>
+    </xs:annotation>
+    <xs:simpleContent>
+      <xs:extension base="tns:nonEmptyStringCollapsedType">
+        <xs:attribute type="xs:int" name="line" />
+        <xs:attribute type="xs:int" name="column" />
+        <xs:attribute type="xs:string" name="file" use="required" />
+        <xs:attribute type="tns:toolDiagnosticSeverityType" name="severity" use="required" />
+        <xs:attribute type="tns:nonEmptyStringType" name="source" />
+      </xs:extension>
```
Comment (dmolineus) [inline doc/report.xsd:110]: Reminder: Adding `lineend` and `columnend` as well

File: src/Command/ExecCommand.php [hunk 3]
```diff
@@ -94,13 +98,16 @@ protected function doExecute(): int
...
-                BufferedOutput::CHANNEL_STRERR
+                BufferedOutput::CHANNEL_STDERR
             );
             $taskOutput->release();
             return (int) $throwable->getCode();
         }
         $taskOutput->release();
 
+        // Fixme: Save/Handle output
+        //$report->asXML(getcwd() . '/' . $this->c->getArtifactOutputPath() . '/checkstyle.xml');
```
Comment (dmolineus) [inline src/Command/ExecCommand.php:109]: I'd use only the `ConsoleWriter` here

File: src/Command/RunCommand.php [hunk 4]
```diff
@@ -41,10 +48,10 @@ protected function configure(): void
             'Define a specific tool which should be run'
         );
         $this->addOption(
-            'keep-going',
-            'k',
+            'fast-finish',
+            'f',
```
Comment (dmolineus) [inline src/Command/RunCommand.php:52]: I don't like the shortcut here, as it's quite often used as `force` options. I suggest using `ff` instead.

File: src/Command/RunCommand.php [hunk 5]
```diff
@@ -85,28 +95,34 @@ protected function doExecute(): int
...
+                    break;
                 }
             }
             $taskOutput->release();
         }
 
+        $report->complete($exitCode === 0 ? Report::STATUS_PASSED : Report::STATUS_FAILED);
+        ReportWriter::writeReport(getcwd() . '/' . $projectConfig->getArtifactOutputPath(), $report);
+        CheckstyleReportWriter::writeReport(getcwd() . '/' . $projectConfig->getArtifactOutputPath(), $report);
+        ConsoleWriter::writeReport($this->output, $report);
```
Comment (dmolineus) [inline src/Command/RunCommand.php:123]: Which report is used should be configurable by the command options but my be handled in a separate PR.

File: doc/report.xsd [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+  <xs:complexType name="reportFile">
+    <xs:sequence>
+      <xs:element name="tools" type="tns:toolListType">
+      </xs:element>
+    </xs:sequence>
+    <xs:attribute name="status" type="tns:reportStatusType" />
+    <xs:attribute name="started_at" type="xs:dateTime"/>
+    <xs:attribute name="completed_at" type="xs:dateTime"/>
+    <xs:attribute name="checkstyleFile" type="xs:string"/>
```
Comment (dmolineus) [inline doc/report.xsd:23]: I'd drop the checkstyle information here as the generated reports should depend on the command configuration

File: src/OutputTransformer/ConsoleOutputTransformerFactory.php [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+                if (OutputInterface::CHANNEL_STDERR === $channel) {
+                    $this->stdErr .= $data;
+                    return;
+                }
+            }
+
+            public function finish(int $exitCode): void
+            {
+                $content = '';
+                while ($line = $this->data->fetch()) {
```
Comment (dmolineus) [inline src/OutputTransformer/ConsoleOutputTransformerFactory.php:67]: There's no need to use the buffered line reader here. Collecting the output in a string parameter would be enough.

File: src/Report/Buffer/ReportBuffer.php [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    public function createToolReport(string $toolName): ToolReportBuffer
+    {
+        $reportName = $toolName;
+        if (isset($this->toolReports[$reportName])) {
+            $number = 0;
+            do {
+                $reportName = $toolName . '-' . ++$number;
+            } while (isset($this->toolReports[$reportName]));
+        }
+        // FIXME: do we rather want to keep the tool name and add the report name as second argument?
```
Comment (dmolineus) [inline src/Report/Buffer/ReportBuffer.php:42]: I prefer keeping the tool name here.

File: src/OutputTransformer/ConsoleOutputTransformerFactory.php (hunk 7 above)
Comment (discordier) [reply src/OutputTransformer/ConsoleOutputTransformerFactory.php:67]: You have in mind that the lines are trimmed by the buffered reader?
This fixes "ugly" line wrapping issues.

Comment (dmolineus) [reply src/OutputTransformer/ConsoleOutputTransformerFactory.php:67]: Missed that point.
